File: packages/jupyterlab-wxyz/src/widgets/yaml.ts
import { safeLoad } from 'js-yaml';

import { JSONModel } from './json';

export class YAMLModel extends JSONModel {
  static model_name = 'YAMLModel';

  defaults() {
    return {
      ...super.defaults(),
      _model_name: YAMLModel.model_name,
      value: {} as any
    };
  }

  theFunction(source: string) {
    return safeLoad(source);
  }
}

File: packages/ipywxyz/src/ipywxyz/widgets/yaml.py
import yaml

from .base import Base, T, W


@W.register
class YAML(Base):
    _model_name = T.Unicode("YAMLModel").tag(sync=True)

    value = T.Union([T.Dict(), T.List(), T.Unicode(), T.Int(), T.Float()]).tag(
        sync=True
    )

    @T.observe("source")
    def _source_changed(self, *_):
        self.value = yaml.safe_load(self.source)
